File: NthNodeFromEnd.java
// Time Complexity : O(n) because we are doing it in one pass
// Space Complexity : O(1) constant space is enough
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach

// here, first we put the fast ans slow at dummy head.. then do fast++ in order to create a gap of 'n'
// once created now we do slow++ and fast++until fast is null. then we change the ptr of next to next.next(deleting)
// and return dummy.head at last

class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        if (head == null)
            return head;

        // so that if head is removed we can handle exception(not losing head0)
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode fast = dummy, slow = dummy;

        for (int i = 0; i < n; i++) {
            fast = fast.next;
        }
        while (fast.next != null) {
            fast = fast.next;
            slow = slow.next;
        }

        slow.next = slow.next.next;

        return dummy.next;

    }
}

File: Nth_node_from_end.py
#time complexity: o(n), traversing all nodes once
#space complexity:o(1)
#passed all cases on LeetCode:yes
#difficulty faced:-
# comments:in the code
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: 'Optional[ListNode]', n: int) -> 'Optional[ListNode]':
        
        #dummy node is introduced for edge case when we have to remove the head
        #the algo will remain the same
        dummy = ListNode(-1)
        slow = fast = dummy 
        dummy.next = head

        count = 0

        #we start fast pointer from n and move till it reaches end
        #this ensures slow ptr starting at start while fast ptr starts at n will goto ()n-i)th pos from start
        #which is what we want to remove 

        #this loop is one way to start fast ptr at nth place
        while count <= n:
            count += 1
            fast = fast.next
        
        while fast!= None:
            fast = fast.next
            slow = slow.next

        slow.next = slow.next.next

        return dummy.next

